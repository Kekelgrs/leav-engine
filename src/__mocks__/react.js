const react = require('react');
module.exports = {...react, memo: x => x};

export default undefined;
