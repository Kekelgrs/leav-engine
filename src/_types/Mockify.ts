// Used to mock any interface, turning all function properties to an optionnal mock
// Mockified object must be then passed to a function with a type assertion
export type Mockify<T> = {[P in keyof T]?: T[P] extends (...args: any) => any ? jest.Mock : T[P]};
