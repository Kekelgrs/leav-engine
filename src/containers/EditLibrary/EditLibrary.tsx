import {History} from 'history';
import * as React from 'react';
import EditLibraryForm from 'src/components/EditLibraryForm';
import Loading from 'src/components/Loading';
import {GET_LIBRARIES_libraries} from 'src/_gqlTypes/GET_LIBRARIES';
import {getLibsQuery, LibrariesQuery} from '../../queries/getLibrariesQuery';
import {SaveLibMutation, saveLibQuery} from '../../queries/saveLibMutation';

interface IEditLibraryProps {
    match: any;
    history: History;
}

class EditLibrary extends React.Component<IEditLibraryProps> {
    constructor(props: IEditLibraryProps) {
        super(props);
    }

    public render() {
        const {match} = this.props;

        const libraryId = match.params.id;
        return libraryId ? (
            <LibrariesQuery query={getLibsQuery} variables={{id: libraryId}}>
                {({loading, error, data}) => {
                    if (loading) {
                        return <Loading />;
                    }
                    if (typeof error !== 'undefined') {
                        return <p>Error: {error.message}</p>;
                    }

                    if (libraryId && !data) {
                        return 'Unknown library';
                    }

                    const libToEdit = data!.libraries !== null && data!.libraries!.length ? data!.libraries![0] : null;

                    return this._getEditLibraryForm(libToEdit);
                }}
            </LibrariesQuery>
        ) : (
            this._getEditLibraryForm(null)
        );
    }

    /**
     * Retrieve EditLibraryForm, wrapped by mutation component
     * @param libToEdit
     * @param history
     */
    private _getEditLibraryForm = (libToEdit: GET_LIBRARIES_libraries | null) => (
        <SaveLibMutation mutation={saveLibQuery}>
            {saveLibrary => {
                const onFormSubmit = async libData => {
                    await saveLibrary({
                        variables: {
                            libData: {
                                id: libData.id,
                                label: {
                                    fr: libData.label.fr,
                                    en: libData.label.en
                                }
                            }
                        },
                        refetchQueries: [{query: getLibsQuery, variables: {id: libData.id}}, {query: getLibsQuery}]
                    });

                    this.props.history.replace({pathname: '/libraries/edit/' + libData.id});
                };
                return <EditLibraryForm library={libToEdit} onSubmit={onFormSubmit} />;
            }}
        </SaveLibMutation>
    )
}

export default EditLibrary;
