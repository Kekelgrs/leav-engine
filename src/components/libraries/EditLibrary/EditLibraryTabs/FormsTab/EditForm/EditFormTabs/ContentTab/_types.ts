import {DragObjectWithType} from 'react-dnd';
import {GET_ATTRIBUTES_attributes_list} from '../../../../../../../../_gqlTypes/GET_ATTRIBUTES';
import {FormElementTypes} from '../../../../../../../../_gqlTypes/globalTypes';
import {IKeyValue} from '../../../../../../../../_types/shared';

export const PLACEHOLDER_ID = '__placeholder__';

export enum UIElementTypes {
    DIVIDER = 'divider',
    FIELDS_CONTAINER = 'fields_container',
    TEXT_BLOCK = 'text_block',
    TABS = 'tabs'
}

export enum FieldTypes {
    TEXT_INPUT = 'input_field',
    DATE = 'date',
    CHECKBOX = 'checkbox',
    ENCRYPTED = 'encrypted',
    DROPDOWN = 'dropdown'
}

export enum DraggableElementTypes {
    RESERVE_LAYOUT_ELEMENT = 'RESERVE_LAYOUT_ELEMENT',
    ATTRIBUTE = 'ATTRIBUTE',
    FORM_ELEMENT = 'FORM_ELEMENT'
}

export interface IFormElement {
    id: string;
    order: number;
    type: FormElementTypes;
    uiElement: IUIElement;
    containerId: string;
    settings?: IKeyValue<any>;
    herited?: boolean;
}

export interface IUIElement {
    type: UIElementTypes | FieldTypes;
    component: JSX.Element;
    canDrop: (dropCandidate: IFormElement) => boolean;
    settings?: string[];
}

export interface IFormElementProps<T extends object> {
    elementData?: IFormElement;
    settings: T;
}

export interface ICommonFieldsSettings {
    label?: string;
    attribute?: GET_ATTRIBUTES_attributes_list;
}

export interface IFormElementPos {
    order: number;
    containerId: string;
}

export interface IFormBuilderDragObject<T extends IUIElement | IFormElement> extends DragObjectWithType {
    type: DraggableElementTypes;
    element: T;
    index: number;
    currentPos?: IFormElementPos;
    originPos?: IFormElementPos;
    dropAtPos?: IFormElementPos;
}
