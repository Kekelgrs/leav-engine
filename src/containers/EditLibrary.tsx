import gql from 'graphql-tag';
import * as React from 'react';
import {Query} from 'react-apollo';
import EditLibraryForm from '../components/EditLibraryForm';
import {SaveLibMutation, saveLibQuery} from '../queries/saveLibMutation';
import {GET_LIBRARY, GET_LIBRARYVariables} from '../_types/GET_LIBRARY';

interface IEditLibraryProps {
    match: any;
}

const libQuery = gql`
    query GET_LIBRARY($id: ID!) {
        libraries(id: $id) {
            id
            system
            label {
                fr
                en
            }
        }
    }
`;

class LibraryQuery extends Query<GET_LIBRARY, GET_LIBRARYVariables> {}

const updateFunc = (cache, {data: {saveLibrary}}) => {
    const {libraries} = cache.readQuery({query: libQuery, variables: {id: saveLibrary.id}});
    cache.writeQuery({
        query: libQuery,
        data: {libraries: libraries.concat([saveLibrary])}
    });
};

function EditLibrary({match}: IEditLibraryProps): JSX.Element {
    const libraryId = match.params.id;
    return (
        <LibraryQuery query={libQuery} variables={{id: libraryId}}>
            {({loading, error, data}) => {
                if (loading || !data) {
                    return <p>Loading</p>;
                }
                if (typeof error !== 'undefined') {
                    return <p>Error: {error.message}</p>;
                }

                return data.libraries !== null && data.libraries.length ? (
                    <SaveLibMutation mutation={saveLibQuery} update={updateFunc}>
                        {saveLibrary => {
                            const onFormSubmit = libData =>
                                saveLibrary({
                                    variables: {
                                        libData: {
                                            id: libData.id,
                                            label: {
                                                fr: libData.label.fr,
                                                en: libData.label.en
                                            }
                                        }
                                    }
                                });
                            return (
                                <EditLibraryForm
                                    library={data.libraries ? data.libraries[0] : null}
                                    onSubmit={onFormSubmit}
                                />
                            );
                        }}
                    </SaveLibMutation>
                ) : (
                    'Unknown library'
                );
            }}
        </LibraryQuery>
    );
}

export default EditLibrary;
