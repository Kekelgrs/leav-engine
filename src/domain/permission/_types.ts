import {
    AppPermissionsActions,
    IPermissionsTreeTarget,
    LibraryPermissionsActions,
    PermissionsActions,
    PermissionTypes
} from '_types/permissions';
import {IQueryInfos} from '_types/queryInfos';
import {ITreeNode} from '_types/tree';

export interface IPermissionTarget {
    attributeId?: string;
    recordId?: string;
}

export interface IGetSimplePermissionsParams {
    type: PermissionTypes;
    applyTo: string | null;
    action: PermissionsActions;
    usersGroupId: string;
    permissionTreeTarget?: IPermissionsTreeTarget;
    ctx: IQueryInfos;
}

export interface IGetPermissionsByActionsParams {
    type: PermissionTypes;
    applyTo: string | null;
    actions: PermissionsActions[];
    usersGroupId: string;
    permissionTreeTarget?: IPermissionsTreeTarget;
    ctx: IQueryInfos;
}

export type PermByActionsRes = {[name: string]: boolean | null} | null;

export interface IGetPermissionByUserGroupsParams {
    type: PermissionTypes;
    action: PermissionsActions;
    userGroupsPaths: ITreeNode[][];
    applyTo?: string;
    permissionTreeTarget?: IPermissionsTreeTarget;
    ctx: IQueryInfos;
}

export interface IGetAppPermissionParams {
    action: AppPermissionsActions;
    userId: string;
    ctx: IQueryInfos;
}

export interface IGetHeritedAppPermissionParams {
    action: AppPermissionsActions;
    userGroupId: string;
    ctx: IQueryInfos;
}

export interface IGetLibraryPermissionParams {
    action: LibraryPermissionsActions;
    libraryId: string;
    userId: string;
    ctx: IQueryInfos;
}

export interface IGetHeritedLibraryPermissionParams {
    action: LibraryPermissionsActions;
    libraryId: string;
    userGroupId: string;
    ctx: IQueryInfos;
}

export interface IGetHeritedPermissionsParams {
    type: PermissionTypes;
    applyTo: string;
    action: PermissionsActions;
    userGroupId: string;
    permissionTreeTarget?: IPermissionsTreeTarget;
    ctx: IQueryInfos;
}

export interface IIsAllowedParams {
    type: PermissionTypes;
    action: PermissionsActions;
    userId: string;
    applyTo?: string;
    target?: IPermissionTarget;
    ctx: IQueryInfos;
}
