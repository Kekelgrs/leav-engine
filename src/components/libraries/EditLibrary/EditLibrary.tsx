import {History} from 'history';
import {i18n} from 'i18next';
import React from 'react';
import {withNamespaces, WithNamespaces} from 'react-i18next';
import useLang from '../../../hooks/useLang';
import useUserData from '../../../hooks/useUserData';
import {getLibsQuery, LibrariesQuery} from '../../../queries/libraries/getLibrariesQuery';
import {SaveLibMutation, saveLibQuery} from '../../../queries/libraries/saveLibMutation';
import {GET_LIBRARIES_libraries} from '../../../_gqlTypes/GET_LIBRARIES';
import {PermissionsActions} from '../../../_gqlTypes/globalTypes';
import Loading from '../../shared/Loading';
import EditLibraryForm from '../EditLibraryForm';

interface IEditLibraryProps extends WithNamespaces {
    match: any;
    history: History;
    i18n: i18n;
}

function EditLibrary({match, history}: IEditLibraryProps) {
    const libraryId = match.params.id;
    const {lang} = useLang();
    const userData = useUserData();
    const readOnly = !userData.permissions[PermissionsActions.admin_edit_library];

    /**
     * Retrieve EditLibraryForm, wrapped by mutation component
     * @param libToEdit
     * @param history
     */
    const _getEditLibraryForm = (libToEdit: GET_LIBRARIES_libraries | null) => (
        <SaveLibMutation mutation={saveLibQuery}>
            {saveLibrary => {
                const onFormSubmit = async libData => {
                    await saveLibrary({
                        variables: {
                            libData: {
                                id: libData.id,
                                label: {
                                    fr: libData.label.fr,
                                    en: libData.label.en
                                },
                                recordIdentityConf:
                                    libData.recordIdentityConf !== null
                                        ? {
                                              label: libData.recordIdentityConf.label,
                                              preview: libData.recordIdentityConf.preview,
                                              color: libData.recordIdentityConf.color
                                          }
                                        : null
                            }
                        },
                        refetchQueries: [{query: getLibsQuery, variables: {id: libData.id}}, {query: getLibsQuery}]
                    });

                    history.replace({pathname: '/libraries/edit/' + libData.id});
                };

                const onPermissionsFormSubmit = async libData => {
                    await saveLibrary({
                        variables: {
                            libData: {
                                id: libData.id,
                                permissionsConf: {
                                    permissionTreeAttributes: libData.permissionsConf.permissionTreeAttributes,
                                    relation: libData.permissionsConf.relation
                                }
                            }
                        },
                        refetchQueries: [{query: getLibsQuery, variables: {id: libData.id}}]
                    });
                };
                return (
                    <EditLibraryForm
                        library={libToEdit}
                        onSubmit={onFormSubmit}
                        onPermsSettingsSubmit={onPermissionsFormSubmit}
                        readOnly={readOnly}
                    />
                );
            }}
        </SaveLibMutation>
    );

    return libraryId ? (
        <LibrariesQuery query={getLibsQuery} variables={{id: libraryId, lang}}>
            {({loading, error, data}) => {
                if (loading) {
                    return <Loading />;
                }
                if (typeof error !== 'undefined') {
                    return <p>Error: {error.message}</p>;
                }

                if (libraryId && !data) {
                    return 'Unknown library';
                }

                const libToEdit = data!.libraries !== null && data!.libraries!.length ? data!.libraries![0] : null;

                return _getEditLibraryForm(libToEdit);
            }}
        </LibrariesQuery>
    ) : (
        _getEditLibraryForm(null)
    );
}

export default withNamespaces()(EditLibrary);
