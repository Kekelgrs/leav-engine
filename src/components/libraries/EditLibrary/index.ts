import EditLibrary from './EditLibrary';
export default EditLibrary;
