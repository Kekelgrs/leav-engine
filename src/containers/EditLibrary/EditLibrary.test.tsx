import {History} from 'history';
import * as React from 'react';
import {MockedProvider} from 'react-apollo/test-utils';
import {create} from 'react-test-renderer';
import {Mockify} from '../../_types/Mockify';
import EditLibrary from './EditLibrary';

describe('EditLibrary', () => {
    test('Snapshot test', async () => {
        const mockMatch: any = {params: {id: 'test'}};
        const mockHistory: Mockify<History> = {};

        const comp = create(
            <MockedProvider>
                <EditLibrary match={mockMatch} history={mockHistory as History} />
            </MockedProvider>
        );

        expect(comp).toMatchSnapshot();
    });
});
