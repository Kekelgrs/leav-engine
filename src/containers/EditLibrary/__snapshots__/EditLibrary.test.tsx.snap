// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditLibrary Snapshot test 1`] = `
<div
  className="ui active centered inline loader"
>
  
</div>
`;
